components/resource_metering/src/reporter/single_target.rs

```diff
@@ -94,24 +116,93 @@ impl DataSink for SingleTargetDataSink {
                 warn!("failed to receive from a grpc call"; "error" => ?err);
             }
         });
+    }
+
+    fn update_data_sink_and_client(&mut self, new_address: String) {
+        if new_address.is_empty() {
+            self.reset();
+            return;
+        }
+
+        if self.address != new_address {
+            self.address = new_address;
+            let channel = {
+                let cb = ChannelBuilder::new(self.env.clone())
+                    .keepalive_time(Duration::from_secs(10))
+                    .keepalive_timeout(Duration::from_secs(3));
+                cb.connect(&self.address)
+            };
+            self.client = Some(ResourceUsageAgentClient::new(channel));
+        }
+
+        if self.data_sink.is_none() {
+            let data_sink = Box::new(DataSinkImpl {
+                scheduler: self.scheduler.clone(),
+            });
+            self.data_sink = Some(self.data_sink_reg.register(data_sink));
+        }
+    }
 
+    fn reset(&mut self) {
+        self.data_sink = None;
+        self.client = None;
```

Comment:
I'm not familiar with this. Seems that here we simply drop it. Is it fine and will there be resource leaks?

Reply:
It's fine. Dropping the inner channel will release resources.